api/v1beta1/azuremanagedcontrolplane_webhook.go

```diff
@@ -440,12 +440,48 @@ func validateLoadBalancerProfile(loadBalancerProfile *LoadBalancerProfile, fldPa
 }
 
 // validateAPIServerAccessProfile validates an APIServerAccessProfile.
-func (m *AzureManagedControlPlane) validateAPIServerAccessProfile(_ client.Client) field.ErrorList {
-	if m.Spec.APIServerAccessProfile != nil {
+func validateAPIServerAccessProfile(apiServerAccessProfile *APIServerAccessProfile, fldPath *field.Path) field.ErrorList {
+	if apiServerAccessProfile != nil {
 		var allErrs field.ErrorList
-		for _, ipRange := range m.Spec.APIServerAccessProfile.AuthorizedIPRanges {
+		for _, ipRange := range apiServerAccessProfile.AuthorizedIPRanges {
 			if _, _, err := net.ParseCIDR(ipRange); err != nil {
-				allErrs = append(allErrs, field.Invalid(field.NewPath("Spec", "APIServerAccessProfile", "AuthorizedIPRanges"), ipRange, "invalid CIDR format"))
+				allErrs = append(allErrs, field.Invalid(fldPath, ipRange, "invalid CIDR format"))
+			}
+		}
+
+		// privateDNSZone should either be "System" or "None" or the resource ID with ending "private.<region>.azmk8s.io"
+		// or ending with "<subzone>.private.<region>.azmk8s.io". The validation below follows the guidelines at
+		// https://learn.microsoft.com/en-us/azure/aks/private-clusters?tabs=azure-portal#configure-a-private-dns-zone
+		if apiServerAccessProfile.PrivateDNSZone != nil {
+			privateDNSZone := *apiServerAccessProfile.PrivateDNSZone
+
+			// Each label of the private dns zone must only contain letters, numbers, underscores, and/or dashes.
+			// And it should be separated from other labels by a period.
+			privateDNSwithZone := `^/subscriptions/[0-9a-fA-F-]+/resourceGroups/[a-zA-Z0-9_-]+/providers/Microsoft\.Network/privateDnsZones/privatelink\.[a-zA-Z0-9_-]+\.(azmk8s\.io)$`
+			regionRegex := regexp.MustCompile(privateDNSwithZone)
+
+			privateDNSwithSubzone := `^/subscriptions/[0-9a-fA-F-]+/resourceGroups/[a-zA-Z0-9_-]+/providers/Microsoft\.Network/privateDnsZones/[a-zA-Z0-9_-]+\.privatelink\.[a-zA-Z0-9_-]+\.(azmk8s\.io)$`
+			subzoneRegex := regexp.MustCompile(privateDNSwithSubzone)
+			if !strings.EqualFold(strings.ToLower(privateDNSZone), "system") && !strings.EqualFold(strings.ToLower(privateDNSZone), "none") {
+				// check if privateDNSZone is a valid resource ID
+				if !regionRegex.MatchString(privateDNSZone) && !subzoneRegex.MatchString(privateDNSZone) {
+					allErrs = append(allErrs, field.Invalid(fldPath, privateDNSZone, "invalid private DNS zone"))
+				}
+
+				// check if subzone is less than 32 characters
+				subzone := `/([^/.]+)\.privatelink\.[^/]+\.azmk8s\.io$`
+				subzoneRegex = regexp.MustCompile(subzone)
+				if subzoneRegex.MatchString(privateDNSZone) {
+					matches := subzoneRegex.FindStringSubmatch(privateDNSZone)
+					fmt.Println(matches)
+					if len(matches) > 0 {
+						subzone := matches[1]
+						fmt.Println(subzone)
```

Comment:
Is this a print statement left over from testing?

Reply:
Oh yes. I will remove them.

Reply:
Added via b0a9d01ea79c2b42bf0a55ae4627dd55fffe5dce